### PR TITLE
Add vfat support for the filesystem module

### PR DESCRIPTION
**Pull request by kwohlfahrt**
##### SUMMARY
This adds vfat support to the filesystem module. In my case, I need this because EFI partitions must be FAT32 formatted.

I was also quite confused by the logic in the module, so I re-factored it to make it (IMHO) clearer. This is a separate commit so it can be dropped/split into another PR if it is controversial.

##### ISSUE TYPE
 - Feature Pull Request

##### COMPONENT NAME
filesystem module

##### ANSIBLE VERSION
```
ansible 2.4.0
  config file = /etc/ansible/ansible.cfg
  configured module search path = [u'/home/kjw53/.ansible/plugins/modules', u'/usr/share/ansible/plugins/modules']
  ansible python module location = /home/kjw53/.pip/lib/python2.7/site-packages/ansible-2.4.0-py2.7.egg/ansible
  executable location = /home/kjw53/.pip/bin/ansible
  python version = 2.7.13 (default, Jan 19 2017, 14:48:08) [GCC 6.3.0 20170118]
```

##### Additional Information

I can't seem to run tests on Debian Stretch (pytest does not recognize the `--boxed` option). If I can somehow add tests to the filesystem module, I am happy to, just can't find any documentation on the subject.

### CODE REVIEW

**Comment by pchauncey** [comment]
This and https://github.com/ansible/ansible/pull/23719 are quite similar.  @abulimov any preference between these?

**Comment by pchauncey** [comment]
Here are some results of testing with ansible-2.4.0:

Initial create and force both work well. 
However, I'm getting this error on resize attempt against an LVM device:
```
ansible localhost -m filesystem -a "dev=/dev/fedora/fat-test fstype=vfat resizefs=true" -b                                                                                      ~/git/ansible
localhost | FAILED! => {
    "changed": false, 
    "failed": true, 
    "module_stderr": "Shared connection to localhost closed.\r\n", 
    "module_stdout": "Traceback (most recent call last):\r\n  File \"/tmp/ansible_6n3pKT/ansible_module_filesystem.py\", line 268, in <module>\r\n    main()\r\n  File \"/tmp/ansible_6n3pKT/ansible_module_filesystem.py\", line 248, in main\r\n    fssize_in_bytes = _get_fs_size(fssize_cmd, dev, module)\r\n  File \"/tmp/ansible_6n3pKT/ansible_module_filesystem.py\", line 131, in _get_fs_size\r\n    return block_size*block_count\r\nUnboundLocalError: local variable 'block_size' referenced before assignment\r\n", 
    "msg": "MODULE FAILURE", 
    "rc": 0
}
```

**Comment by kwohlfahrt** [comment]
This should be fixed. Note there is a bug that XFS resize always reports changed, even if it has not changed. This is present on `ansible/devel` as well, so not a regression in this PR.

**Comment by pchauncey** [comment]
@kwohlfahrt   Regarding pytest --boxed, this worked for me:
```
git clone https://github.com/pytest-dev/pytest-boxed
pip install -e pytest-boxed
```

Regarding testing this module, for me everything works except resizing.  Have you been able to test resizes with the module?  Here are some of the issues I'm running into:

```
$ sudo lvresize -L +100M /dev/fedora/fat-test
  Size of logical volume fedora/fat-test changed from 500.00 MiB (125 extents) to 600.00 MiB (150 extents).
  Logical volume fedora/fat-test successfully resized.
$ ansible localhost -m filesystem -a "dev=/dev/fedora/fat-test fstype=vfat resizefs=true" -b                                                                                               
localhost | FAILED! => {
    "changed": false, 
    "err": "You must specify new size.\n", 
    "failed": true, 
    "msg": "Resizing filesystem vfat on device '/dev/fedora/fat-test' failed", 
    "rc": 1
}
```

On a USB stick that I've changed the partition size of the vfat partition, I'm getting this error:
```
localhost | FAILED! => {
    "changed": false, 
    "err": "Warning: The driver descriptor says the physical block size is 2048 bytes, but Linux says it is 512 bytes.\n", 
    "failed": true, 
    "msg": "Failed to get block count and block size of /dev/sdd1 with /sbin/fatresize", 
    "rc": 1
}
```

Any ideas?

**Comment by kwohlfahrt** [comment]
The first issue should now be fixed. The second, I can't reproduce. Could you let me know how you partitioned the USB? Or `dd` the offending stick and post it somewhere if it's small enough?

File: lib/ansible/modules/system/filesystem.py
Comment (pchauncey) [inline lib/ansible/modules/system/filesystem.py]: Running into some issues here: 

1. Getting an `UnboundLocalError: local variable devsize_in_bytes referenced before assignment`. To reproduce, either `force=true` or zero out the device first.
2. Effectively this is executing: `blkid -s devsize_in_bytes /dev/blah`, which doesn't seem intended.  Did you intend to run `mkfs` here instead of `cmd`?
3. I don't think vfat needs special attention here at all.

File: lib/ansible/modules/system/filesystem.py
Comment (pchauncey) [inline lib/ansible/modules/system/filesystem.py]: This should probably be `mkfs` also instead of `cmd`.  Additionally, we've lost the `force_flag` and `opts` for some reason.   

How does this look to you: 
```diff
@@ -232,10 +232,7 @@ def main():
             force_flag = ''
         if opts is None:
             opts = ''
-        if fstype == 'vfat':
-            rc,out,err = module.run_command("%s -s %d %s" % (cmd, devsize_in_bytes, dev))
-        else:
-            rc,out,err = module.run_command("%s %s" % (cmd, dev))
+        rc,out,err = module.run_command("%s %s %s '%s'" % (mkfs, force_flag, opts, dev))
         if rc == 0:
             module.exit_json(changed=True)
         else:
```

**Review by pchauncey — CHANGES_REQUESTED** [review]

File: lib/ansible/modules/system/filesystem.py
Comment (kwohlfahrt) [reply lib/ansible/modules/system/filesystem.py]: You're completely right, this is not correct at all. I wasn't careful rebasing and the changes from the 'resize' section got propagated to here.

**Comment by kwohlfahrt** [comment]
Sorry about the mess in the previous batch. I wasn't careful rebasing and copied the change to the wrong location.

**Review by pchauncey — APPROVED** [review]

**Comment by pchauncey** [comment]
@kwohlfahrt thanks for all of the updates.  Tested with ansible 2.4.0.

**Comment by kwohlfahrt** [comment]
No changes, just rebased onto latest devel.

**Comment by kwohlfahrt** [comment]
Again, rebased onto latest devel. @pchauncey can this be merged? Or do I need to say:

ready_for_review

**Comment by kwohlfahrt** [comment]
Due to the recent refactor of the filesystem module, I'm re-writing this enhancement.

**Comment by pilou-** [comment]
@kwohlfahrt thanks !

**Comment by mattclay** [comment]
CI failure in integration tests:

https://app.shippable.com/github/ansible/ansible/runs/50185/38/tests

**Comment by kwohlfahrt** [comment]
Just FreeBSD failing now, due to the command not being `mkfs.vfat` but `newfs_msdos` instead. What's the best way to detect the current system and override this?

File: lib/ansible/modules/system/filesystem.py
Comment (pilou-) [inline lib/ansible/modules/system/filesystem.py]: Community module freeze is 7 February 2018, it might reach 2.5: use 2.5 instead of 2.6.

File: lib/ansible/modules/system/filesystem.py
Comment (pilou-) [inline lib/ansible/modules/system/filesystem.py]: Isn't `dev` missing in this command ?

File: lib/ansible/modules/system/filesystem.py
Comment (pilou-) [inline lib/ansible/modules/system/filesystem.py]: Instead of overriding `Filesystem.grow` should not a new method `get_grow_command(self, dev, devsize_in_bytes)` be added ? Then this method would be called in `Filesystem.grow` and the default implementation in `Filesystem` would return `"%s %s" % (self.module.get_bin_path(self.GROW, required=True), dev)`.

File: test/integration/targets/filesystem/tasks/overwrite_another_fs.yml
Comment (pilou-) [inline test/integration/targets/filesystem/tasks/overwrite_another_fs.yml]: The purpose of this task is to create a filesystem which isn't handled by the filesystem module. Once created we try to create another filesystem (without using `force`) supported by the filesystem module, then this task must fail. This task should not depend on `fstype` value.

Creating an UFS filesystem on FreeBSD is ok but creating an ext2 filesystem isn't because `ext2` filesystem is supported by `filesystem` module.

File: test/integration/targets/filesystem/defaults/main.yml
Comment (pilou-) [inline test/integration/targets/filesystem/defaults/main.yml]: Which test systems provide fatresize < 1.0.4 ?

File: test/integration/targets/filesystem/defaults/main.yml
Comment (pilou-) [inline test/integration/targets/filesystem/defaults/main.yml]: Is `-F12` parameter mandatory ?

File: lib/ansible/modules/system/filesystem.py
Comment (kwohlfahrt) [reply lib/ansible/modules/system/filesystem.py]: I didn't do it this way since the `devsize_in_bytes` parameter would be ignored by the other classes, so its a bit misleading to have that.

File: test/integration/targets/filesystem/tasks/overwrite_another_fs.yml
Comment (kwohlfahrt) [reply test/integration/targets/filesystem/tasks/overwrite_another_fs.yml]: Ah, right. I thought it just tests that it doesn't overwrite any existing filesystem (which is a necessary test too). I can change it to another Linux FS, but can't think of any common ones that the module doesn't handle.

File: test/integration/targets/filesystem/defaults/main.yml
Comment (kwohlfahrt) [reply test/integration/targets/filesystem/defaults/main.yml]: I'm not 100% sure what's necessary on FreeBSD, I was just copying the test command here since it doesn't seem to work without any options. `-F12` is also not enough, it complains about a missing label - I'm not sure why the command in `overwrite_another_fs.yml` works and this doesn't?

File: test/integration/targets/filesystem/defaults/main.yml
Comment (kwohlfahrt) [reply test/integration/targets/filesystem/defaults/main.yml]: On inspection this release is a project on Github, I'm not sure how official it is. I can't get it to work reliably on my system, so I'm going to drop resizing from this PR for now.

File: lib/ansible/modules/system/filesystem.py
Comment (pilou-) [reply lib/ansible/modules/system/filesystem.py]: Instead `devsize` attribute could be set in `grow` (`self.devsize = devsize_in_bytes`) then reused latter or `dev` could become an object with a `size` property.

File: test/integration/targets/filesystem/tasks/overwrite_another_fs.yml
Comment (pilou-) [reply test/integration/targets/filesystem/tasks/overwrite_another_fs.yml]: A swap partition (created using `mkswap`) could be used.

Comment (kwohlfahrt) [reply test/integration/targets/filesystem/tasks/overwrite_another_fs.yml]: `newfs` does not seem to work on image files, so I'll use `mkfs.ext2`.

Like you say it is different in that it now tests a filesystem supported by the module, but it still tests that an existing filesystem is not overwritten which I think is the important part. I don't think the control flow is any different (apart from the fact that `FILESYSTEMS.get` returns a class instead of `None`).

Comment (kwohlfahrt) [reply test/integration/targets/filesystem/tasks/overwrite_another_fs.yml]: Sorry didn't see your previous comment. I'll update it to use a swap partition instead.

File: lib/ansible/modules/system/filesystem.py
Comment (kwohlfahrt) [reply lib/ansible/modules/system/filesystem.py]: I will implement the second option.

File: test/integration/targets/filesystem/tasks/overwrite_another_fs.yml
Comment (kwohlfahrt) [reply test/integration/targets/filesystem/tasks/overwrite_another_fs.yml]: `mkswap` doesn't exist on FreeBSD (and `newfs` doesn't work). Any ideas what to use instead?

File: lib/ansible/modules/system/filesystem.py
Comment (pilou-) [inline lib/ansible/modules/system/filesystem.py]: `self.module` isn't defined (🆗  DONE)

File: lib/ansible/modules/system/filesystem.py
Comment (pilou-) [inline lib/ansible/modules/system/filesystem.py]: s/pqth/path/ (:ok: DONE)

File: lib/ansible/modules/system/filesystem.py
Comment (pilou-) [inline lib/ansible/modules/system/filesystem.py]: In order to avoid to execute command twice, size could be stored using something like:
```
def __init__(self, path):
    self._size = None
[...]
def size(self):
    if self._size is None:
        # compute self._size
        [...]
    return self._size
```

File: test/integration/targets/filesystem/tasks/overwrite_another_fs.yml
Comment (pilou-) [reply test/integration/targets/filesystem/tasks/overwrite_another_fs.yml]: > Ah, right. I thought it just tests that it doesn't overwrite any existing filesystem (which is a necessary test too).

The use case _supported filesystem aren't overwritten when force isn't used_ is already tested [there](https://github.com/kwohlfahrt/ansible/blob/db7513ec2f9721333275e2deafa6c576de2fa526/test/integration/targets/filesystem/tasks/create_fs.yml#L19).

File: lib/ansible/modules/system/filesystem.py
Comment (kwohlfahrt) [reply lib/ansible/modules/system/filesystem.py]: Yes, but the cost is not much - maybe 1ms. I think the simple solution is better for now.

File: test/integration/targets/filesystem/tasks/overwrite_another_fs.yml
Comment (kwohlfahrt) [reply test/integration/targets/filesystem/tasks/overwrite_another_fs.yml]: Ah, I see. Should this test be skipped on FreeBSD then?

Comment (pilou-) [reply test/integration/targets/filesystem/tasks/overwrite_another_fs.yml]: ok (filesystems created with `newfs` aren't recognized by `blkid`)

**Comment by pilou-** [comment]
@kwohlfahrt great ! Could you remove the `when: ansible_system != 'FreeBSD'` condition in `overwrite_another_fs.yml` ?

**Comment by kwohlfahrt** [comment]
That condition is there as discussed [here](https://github.com/ansible/ansible/pull/23527#discussion_r161587667) - FreeBSD does not have `mkswap`, and `newfs` doesn't work on image files so there is no unsupported filesystem to test.

**Comment by pilou-** [comment]
`overwrite_another_fs.yml` is [skipped](https://github.com/ansible/ansible/pull/23527/files#diff-1e19b671068be3d6541492ab23955a93) when `ansible_system != 'FreeBSD'` that's why the same condition in `overwrite_another_fs.yml` should be removed.

**Comment by kwohlfahrt** [comment]
Oh, I see what you mean. Done. Note the skeleton code for `fatresize` is still there, but not enabled or tested as I had some problems with segfaults in this tool.

**Comment by pilou-** [comment]
[There](https://github.com/ansible/ansible/pull/23527/files#diff-bfde8baf1f58b12681d33f80e89c681cR240) it should be `dev.path` instead of `dev`. I will try `fatresize` on some distributions tested by the CI and report result there.

**Comment by kwohlfahrt** [comment]
`str(Device('foo')) == 'foo' == ('%s' % Device('foo'))`, so I think this is OK. I have changed to use `str(dev)` throughout for consistency.

File: lib/ansible/modules/system/filesystem.py
Comment (pilou-) [inline lib/ansible/modules/system/filesystem.py]: `str` must be added around `dev.size()`.

**Comment by pilou-** [comment]
oh, I missed the str method, thanks for pointing that.

I was able to reproduce segfault of fatresize (1.0.2). It appears this issue has been fixed in [1.0.4](https://github.com/ya-mouse/fatresize/pull/7).
I propose to:
- enable grow operation on vfat filesystems
- enable grow integration tests with vfat when fatresize >= 1.0.4
- add a note mentioning that grow on vfat filesystems will likely fail when fatresize < 1.0.4

**Comment by kwohlfahrt** [comment]
I'm not sure how official that fatresize repository is, Arch and Debian both list [this](https://sourceforge.net/projects/fatresize/) as their upstream.

**Comment by kwohlfahrt** [comment]
Ah, Arch at least mentions that upstream has moved to that github repository, so I will do as you suggested. Hopefully debian catches up soon.

**Comment by kwohlfahrt** [comment]
Tests are passing, except 'fatresize' is not installed on CentOS-7 when it should be - can you have a look at where my conditional is wrong?

**Comment by pilou-** [comment]
`ansible_os_family` is equal to `RedHat` for CentOS hosts.

**Comment by kwohlfahrt** [comment]
Fixed, I think this is good to go then!

**Comment by pilou-** [comment]
shipit

**Comment by pilou-** [comment]
Hello! @maxamillion could you please review this pull request?

I am a maintainer of this module, I have already approved this pull-request but the workflow requires another approval (from another module maintainer or a system namespace maintainer or a core team member).

The `stale_ci` label should be removed by the bot after the next approval (meaning another run of the CI should be triggered).

**Review by maxamillion — APPROVED** [review]

**Comment by maxamillion** [comment]
shipit

**Comment by pilou-** [comment]
It seems there is a problem with this pull request. The GitHub web interface states that
> This branch has no conflicts with the base branch

then I asked @mscherer  if he could merge this pull-request (this pull-request is labeled with `shipit`), he replied that the GithHub web interface states that
> This branch cannot be rebased safely

What should be done in order to merge this pull-request ?

**Comment by kwohlfahrt** [comment]
I have rebased onto the latest devel locally (without any conflicts). Would it be helpful to push this?

**Comment by pilou-** [comment]
@kwohlfahrt let's try that :)

**Comment by mscherer** [comment]
I can however squash and merge if that's what we want.

**Comment by pilou-** [comment]
@misc `squash` is what should be done according to the [committers guidelines](http://docs.ansible.com/ansible/devel/committer_guidelines.html#general-rules):
> * Do
>   * Squash, avoid merges whenever possible, use github’s squash commits or cherry pick if needed (bisect thanks you).

**Comment by pilou-** [comment]
@kwohlfahrt thank you for this contribution and the hard work.

**Comment by pilou-** [comment]
@kwohlfahrt thank you for this contribution and the hard work.

**Pull request merged**
